boards: mps3: Add support for corstone300/an552 and corstone310/an555

wearyzen opened the pull request:
What is changed?
 - Added a new mps3 board an552 for the soc corstone300. an552 uses ARM FVP (FVP_Corstone_SSE-300_Ethos-U55).  The qualifier to build/run application with board mps3/an552 is
  `mps3/corstone300/an552` for secure and `mps3/corstone300/an552/ns` for non-secure.
 - Added a new mps3 board an555 for the soc corstone310 to enable testing with ARM FVP (FVP_Corstone_SSE-310).
    The qualifier to build/run application with board mps3/an555 is `mps3/corstone310/an555` for secure and     `mps3/corstone310/an555/ns` for non-secure.
 - MPS3 board documentation was modified to include all 3 boards.
 - Note: the qualifier to build/run application with board mps3/an547 is now changed to `mps3/corstone300/an547` for secure and `mps3/corstone300/an547/ns` for non-secure.

How is it changed?
 - Extracted common code for mps3/an547 and mps3/an552 as corstone300.
 - Renamed soc for an547 to corstone300 and added a new soc corstone300/an552 and corstone310/an555.

Why do we need this change?
 - This enables FVP support and testing for mps3/an552 and mps3/an555.
 - SOC/qualifier for mps3/an547 was renamed to reduce code redundancy.

Signed-off-by: Sudan Landge <sudan.landge@arm.com>

wearyzen committed:
boards: mps3: Add support for corstone300/an552

What is changed?
 - Added a new mps3 board an552 for the soc corstone300.
   The qualifier to build/run application with board mps3/an552 is
   `mps3/corstone300/an552` for secure and
   `mps3/corstone300/an552/ns` for non-secure.
 - Added FVP variant to enable FVP testing with corstone300
   and it uses the ARM FVP `FVP_Corstone_SSE-300_Ethos-U55`.
   The qualifier to build/run application with FVP is
   `mps3/corstone300/fvp` for secure and
   `mps3/corstone300/fvp/ns` for non-secure.
 - Note: the qualifier to build/run application with board mps3/an547
   is now changed to
   `mps3/corstone300/an547` for secure and
   `mps3/corstone300/an547/ns` for non-secure.

How is it changed?
 - Moved common code from mps3/an547 to corstone300.
 - Renamed soc for an547 to corstone300 and added
   a new soc corstone300/an552.

Why do we need this change?
 - This enables FVP support and testing for corstone300.
 - SOC/qualifier for mps3/an547 was renamed to reduce code redundancy
 - A separate FVP variant was added for AN552 because, the TFM board
   used for non-secure variant differs for FPGA and FVP.
   TFM board `arm/mps3/corstone300/fvp` should be used when testing
   AN552 with FVP and `arm/mps3/corstone300/an552` should be used when
   testing with AN552 FPGA.

Signed-off-by: Sudan Landge <sudan.landge@arm.com>

wearyzen committed:
boards: mps3: Add support for corstone310/an555

What is changed?
 - Added a new mps3 board an555 for the soc corstone310.
   The qualifier to build/run application with board mps3/an555 is
   `mps3/corstone310/an555` for secure and
   `mps3/corstone310/an555/ns` for non-secure.
 - Added FVP variant to enable FVP testing with corstone310
   and it uses the ARM FVP `FVP_Corstone_SSE-310`.
   The qualifier to build/run application with FVP is
   `mps3/corstone310/an555fvp` for secure and
   `mps3/corstone310/an555fvp/ns` for non-secure.

Why do we need this change?
 - This enables FVP support and testing for corstone310.
 - A separate FVP variant was added for AN555 because, the TFM board
   used for non-secure variant differs for FPGA and FVP.
   TFM board `arm/mps3/corstone310/an555` should be used when testing
   AN555 with FVP and `arm/mps3/corstone310/fvp` should be used when
   testing with AN555 FPGA.

Signed-off-by: Sudan Landge <sudan.landge@arm.com>

dcpleung commented:
Could you also add a note in the next release note about the renaming of `mps3/an547` to `mps3/corstone300/547`?

kartben commented on boards/arm/mps3/doc/mps3_an547.rst:
Please keep the original index.rst file and rework all the variants into the same document rather than vastly duplicating content in the two new pages you introduced. You can probably use a mix of Sphinx "tabs" for the sections that differ per variant, and simply reword paragraphs to refer to "MPS3" when content applies to all variants.

wearyzen replied:
Hi @kartben, thanks for the review. I will try to create the different files as suggested but I had a query regarding the index.rst file. Since we have a `boards/arm/index.rst` which includes all rst files from subdirectories do we still need to keep a file with name index.rst because initially I had referred the `boards/arm/mps2/doc` for my changes and it didn't have one.

kartben commented on boards/arm/mps3/doc/mps3_an552.rst:
```diff
@@ -0,0 +1,240 @@
+Serial Port
+===========
+
+The MPS3 AN552 has six UARTs. The Zephyr console output by default, uses
```
When reworking into a single document, instances like this can probably be worded as:

```suggestion
The MPS3 has six UARTs. The Zephyr console output by default, uses
```

wearyzen commented:
Few more changes were identified so converting this to a draft.

wearyzen committed:
arch: arm: fix null pointer dereference check test

What is changed?
Updated the condition thats prevents mpu config for null dereference.
Added a new check so that mpu is configured for null dereference if
devicetree contains a memory-region node with:
 - node address starting at 0
 - size covered by the node is more than the null dereference page
   size (0x400) and
 - contains a memory-attr

Why is the change needed?
The check relied on flash base address to align with 0 for
configuring the mpu for null dereference but, a device tree
could have a flash starting at an address other than 0 and
still need the mpu config for null dereference.
The new extra check provides a way to connfigure mpu for
null dereference even if flash base address is not 0.

Note, though this change helps with mpu config for new boards having
flash address other than 0, this change does not change existing
behaviour for existing boards.

Signed-off-by: Sudan Landge <sudan.landge@arm.com>

kartben reviewed (dismissed):
Great job with the docs, thank you!

carlescufi commented:
@wearyzen please rebase to resolve conflicts

wearyzen commented:
Code rebased
